lib/nrf_modem_lib/nrf91_net_l2/nrf91_net_l2.c

```diff
@@ -0,0 +1,329 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ *
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/types.h>
+#include <zephyr/net/offloaded_netdev.h>
+#include <zephyr/net/net_l2_connectivity.h>
+#include <zephyr/logging/log.h>
+#include <modem/nrf_modem_lib.h>
+#include <modem/lte_lc.h>
+#include <modem/pdn.h>
+
+#include "ip_addr_helper.h"
+
+LOG_MODULE_REGISTER(nrf91_net_l2, CONFIG_NRF91_NET_L2_LOG_LEVEL);
+
+/* This library don't support the CONFIG_NRF_MODEM_LIB_SYS_INIT option.
+ *
+ * If this option is set, initialization of the nRF Modem library occurs
+ * automatically, at the same time as the Zephyr network stack initializes the nRF91 NET L2 layer.
+ * At SYS INIT - POST_KERNEL init level.
+ *
+ * Due to this we don't support the aforementioned option when building with this library.
+ * We need to control the initialization order to prevent a potential race in case SYS INIT level
+ * and priority changes.
+ */
+BUILD_ASSERT(!IS_ENABLED(CONFIG_NRF_MODEM_LIB_SYS_INIT),
+	     "System initialization of nRF Modem library is not supported");
+
+/* This library requires that Zephyr's IPv4 and IPv6 stacks are enabled.
+ *
+ * The modem supports both IPv6 and IPv4. At any time, either IP families can be given by
+ * the network, independently of each other. Because of this we require that the corresponding
+ * IP stacks are enabled in Zephyr NET in order to add and remove both IP family address types
+ * to the network interface.
+ */
+BUILD_ASSERT(IS_ENABLED(CONFIG_NET_IPV6) && IS_ENABLED(CONFIG_NET_IPV4), "IPv6 and IPv4 required");
+
+/* Forward declarations */
+static void lte_timeout_work_fn(struct k_work *work);
+void pdn_event_handler(uint8_t cid, enum pdn_event event, int reason);
+int nrf91_net_l2_disconnect(const struct net_if_conn *if_conn);
+
+/* Delayable work used to handle LTE connection timeouts. */
+static K_WORK_DELAYABLE_DEFINE(lte_timeout_work, lte_timeout_work_fn);
+
+/* Local reference to the network interface the l2 layer is bound to. */
+static struct net_if *iface_bound;
+
+/* Local functions */
+
+static void lte_timeout_work_fn(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	LOG_DBG("LTE connection timeout");
+
+	int ret = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_DEACTIVATE_LTE);
+
+	if (ret) {
+		LOG_ERR("lte_lc_func_mode_set, error: %d", ret);
+		net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+
+		/* Notify that the connection attempt timed out before returning. */
+	}
+
+	net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_TIMEOUT, iface_bound);
+}
+
+static int modem_init(void)
+{
+	LOG_DBG("Initializing nRF Modem Library");
+
+	int ret = nrf_modem_lib_init(NORMAL_MODE);
+
+	/* Handle return values related to modem DFU. */
+	switch (ret) {
+	case 0:
+		/* Initialization successful, no action required. */
+		return 0;
+	case NRF_MODEM_DFU_RESULT_OK:
+		LOG_DBG("Modem DFU successful. The modem will run the updated "
+			"firmware after reinitialization.");
+		break;
```

Comment:
Is the reinitialization happening on the the subsequent call to `nrf_modem_lib_init()` also if the modem is not shut down first?

Reply:
Yes, once the result of the update is returned by `nrf_modem_init`, the user must call `nrf_modem_shutdown` and then again `nrf_modem_init`. That's documented [here](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/nrf_modem/doc/delta_dfu.html#reinitializing-the-modem-to-run-the-new-firmware).

Reply:
Testing suggested that in order to properly reinitialize the modem after DFU you have to:
shutdown -> init -> shutdown -> init

Reply:
From my tests it seems you need to shutdown once, then init twice; the first time to receive the DFU result, and the second to run the firmware. Calling shutdown after the first init seems has returned a DFU code seems to fail.

Reply:
Interesting, Ill test again

Reply:
You are right, It is not nessecary to shutdown in between calls to init.